open_rarity/resolver/rarity_providers/external_rarity_provider.py

```diff
@@ -277,8 +133,41 @@ def _add_trait_sniper_rarity_data(
         # We're currently using opensea slug to calculate trait sniper slug
         slug = collection_with_metadata.opensea_slug
         if cache_external_ranks:
+            # Try to load cache with a local cache file
             self._load_cache_from_file(slug=slug, rank_provider=rank_provider)
 
+        # If we didn't want to load cache or cache is empty, pull data from API
+        # in bulk and store it in cache
+        if not self._is_cache_loaded(slug=slug, rank_provider=rank_provider):
+            contract_addresses = collection_with_metadata.contract_addresses
+            assert (len(contract_addresses)) == 1
+            try:
+                trait_sniper_rank_data = TraitSniperResolver.get_all_ranks(
+                    contract_address=contract_addresses[0]
+                )
+                logger.debug(
+                    f"Resolved trait sniper rarity for {slug=} {contract_addresses[0]}"
+                )
+            except Exception:
+                logger.exception(
+                    "Failed to resolve Traits Sniper ranking data",
+                    exc_info=True,
+                )
+                return tokens_with_rarity
+            if not trait_sniper_rank_data:
+                logger.warning(f"[TraitSniper] Did not get any data for {slug}")
+                return tokens_with_rarity
+
+            # Load cache with fetched data
+            rank_cache = self._get_provider_rank_cache(
+                slug=slug, rank_provider=rank_provider
+            )
+            for ts_rank_data in trait_sniper_rank_data:
+                token_id = str(ts_rank_data["token_id"])
+                rank_cache[token_id] = int(ts_rank_data["rarity_rank"])
+
+            self.write_cache_to_file(slug, rank_provider)
```

Comment:
Why is `write_cache_to_file` called for trait sniper and rarity sniffer in this file but not for rarity sniper?

It also gets called again in the testset resolver script.

Reply:
Rarity sniper writes cache one token at a time vs. all at once (trait sniper/rarity sniffer). Didn't want to add complex logic to read partial caches.

It's true it's duplicative though - I can remove the 2nd call in testset resolver script.